docker/elasticsearch-setup/create-indices.sh

```diff
@@ -14,22 +14,102 @@ else
+    else
+      concrete_index_name=$(curl -XGET "$ELASTICSEARCH_PROTOCOL://$ELASTICSEARCH_HOST_URL:$ELASTICSEARCH_PORT/_alias/$source_index" | jq 'keys[]' | head -1 | tr -d \")
+      curl -XDELETE "$ELASTICSEARCH_PROTOCOL://$ELASTICSEARCH_HOST_URL:$ELASTICSEARCH_PORT/$concrete_index_name"
+    fi
+
+    curl -XPOST "$ELASTICSEARCH_PROTOCOL://$ELASTICSEARCH_HOST_URL:$ELASTICSEARCH_PORT/_aliases" -H 'Content-Type: application/json' \
+      -d "{\"actions\":[{\"remove\":{\"index\":\"*\",\"alias\":\"$source_index\"}},{\"add\":{\"index\":\"$target_index\",\"alias\":\"$source_index\"}}]}"
+
+    echo -e "\nReindexing to $target_index is complete"
+    return 0
+  else
+    echo -e "\nReindexing to $target_index is failure"
+    return 1
+  fi
+}
+
+function create_index() {
+  generate_index_file "index/$2" "index/$3" /tmp/data
+
+  #checking if index(or alias) exists
+  if [ $(curl -o /dev/null -s -w "%{http_code}" "$ELASTICSEARCH_PROTOCOL://$ELASTICSEARCH_HOST_URL:$ELASTICSEARCH_PORT/$1") -eq 404 ]
+  then
+    echo -e '\ncreating index' "$1"
+
+    curl -XPUT "$ELASTICSEARCH_PROTOCOL://$ELASTICSEARCH_HOST_URL:$ELASTICSEARCH_PORT/$1" -H 'Content-Type: application/json' --data @/tmp/data
+    return 0
+  else
+    echo -e '\ncomparing with existing version of index' "$1"
+
+    curl -XGET "$ELASTICSEARCH_PROTOCOL://$ELASTICSEARCH_HOST_URL:$ELASTICSEARCH_PORT/$1/_settings" | jq '.. | .settings? | select(. != null)' | \
+      jq '. | {"index":{"max_ngram_diff": .index.max_ngram_diff, "analysis": .index.analysis}}' > /tmp/existing_setting
+    curl -XGET "$ELASTICSEARCH_PROTOCOL://$ELASTICSEARCH_HOST_URL:$ELASTICSEARCH_PORT/$1/_mapping" | jq '.. | .mappings? | select(. != null)' > /tmp/existing_mapping
+
+    generate_index_file /tmp/existing_setting /tmp/existing_mapping /tmp/existing
+
+    if diff <(jq -S . /tmp/existing) <(jq -S . /tmp/data)
+    then
+      echo -e "\nexisting version of index $1 is latest"
+      return 0
+    else
+      echo -e "\nupdating index" "$1"
 
-  curl -XPUT $ELASTICSEARCH_PROTOCOL://$ELASTICSEARCH_HOST_URL:$ELASTICSEARCH_PORT/$1 -H 'Content-Type: application/json' --data @/tmp/data
+      reindex "$1" && return 0 || return 1
+    fi
+  fi
 }
 
-create_index chartdocument chart/settings.json chart/mappings.json
-create_index corpuserinfodocument corp-user/settings.json corp-user/mappings.json
-create_index dashboarddocument dashboard/settings.json dashboard/mappings.json
-create_index datajobdocument datajob/settings.json datajob/mappings.json
-create_index dataflowdocument dataflow/settings.json dataflow/mappings.json
-create_index dataprocessdocument data-process/settings.json data-process/mappings.json
-create_index datasetdocument dataset/settings.json dataset/mappings.json
-create_index mlmodeldocument ml-model/settings.json ml-model/mappings.json
-create_index tagdocument tags/settings.json tags/mappings.json
+create_index chartdocument chart/settings.json chart/mappings.json || exit 1
```

Comment:
Are we adding these so that one failure would fail the whole script?

Reply:
Yes, that's the intention.